ChaosMod/Effects/db/Misc/MiscPhoneCall.cpp

```diff
@@ -0,0 +1,356 @@
+};
+
+static uint8_t *lockCellphoneGlobal;
+static CellPhone *cellphoneStateGlobal;
+
+static int phoneScaleform                              = 0;
+static int currentMode                                 = 0;
+static int lastTick                                    = 0;
+static int soundId                                     = 0;
+
+static bool isDialing                                  = true;
+static bool lockCellphone                              = true;
+
+std::unordered_map<Hash, const char *> phoneScaleforms = { { 0xD7114C9, "cellphone_ifruit" },
+	                                                       { 0x9B22DBAF, "cellphone_badger" },
+	                                                       { 0x9B810FA2, "cellphone_facade" } };
+
+static void DrawPhone(PhoneCall &phoneCall)
+{
+	if (isDialing)
+	{
+		BEGIN_SCALEFORM_MOVIE_METHOD(phoneScaleform, "SET_SOFT_KEYS");
+		SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(2);
+		SCALEFORM_MOVIE_METHOD_ADD_PARAM_BOOL(true);
+		SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(5);
+		END_SCALEFORM_MOVIE_METHOD();
+	}
+	else
+	{
+		BEGIN_SCALEFORM_MOVIE_METHOD(phoneScaleform, "SET_SOFT_KEYS");
+		SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(2);
+		SCALEFORM_MOVIE_METHOD_ADD_PARAM_BOOL(true);
+		SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(1);
+		END_SCALEFORM_MOVIE_METHOD();
+	}
+
+	BEGIN_SCALEFORM_MOVIE_METHOD(phoneScaleform, "SET_SOFT_KEYS");
+	SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(3);
+	SCALEFORM_MOVIE_METHOD_ADD_PARAM_BOOL(true);
+	SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(1);
+	END_SCALEFORM_MOVIE_METHOD();
+
+	BEGIN_SCALEFORM_MOVIE_METHOD(phoneScaleform, "SET_DATA_SLOT");
+	SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(4);
+	SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(0);
+	SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(3);
+
+	BEGIN_TEXT_COMMAND_SCALEFORM_STRING(phoneCall.contactName);
+	END_TEXT_COMMAND_SCALEFORM_STRING();
+
+	BEGIN_TEXT_COMMAND_SCALEFORM_STRING("CELL_2000");
+	ADD_TEXT_COMPONENT_SUBSTRING_PLAYER_NAME(phoneCall.profileTxdName);
+	END_TEXT_COMMAND_SCALEFORM_STRING();
+
+	BEGIN_TEXT_COMMAND_SCALEFORM_STRING(isDialing ? "CELL_217" : "CELL_219");
+	END_TEXT_COMMAND_SCALEFORM_STRING();
+
+	END_SCALEFORM_MOVIE_METHOD();
+
+	BEGIN_SCALEFORM_MOVIE_METHOD(phoneScaleform, "DISPLAY_VIEW");
+	SCALEFORM_MOVIE_METHOD_ADD_PARAM_INT(4);
+	END_SCALEFORM_MOVIE_METHOD();
+}
+
+static void UpdatePhone(PhoneCall &phoneCall)
+{
+	DrawPhone(phoneCall);
+	*lockCellphoneGlobal = (uint8_t)lockCellphone; // disables phone inputs
+	TERMINATE_ALL_SCRIPTS_WITH_THIS_NAME("cellphone_controller");
+}
+
+static void PlayPhoneCall(PhoneCall &phoneCall)
+{
+	REGISTER_GLOBAL("CellPhoneState", "dialogue_handler",
+	                "2D 00 02 00 ? 52 ? ? 41 ? 72 08 2A 06 56 ? ? 52 ? ? 41 ? 71 08 20 56 ? ? 72", 6,
+	                GlobalPatternIdiom::GLOBAL_U16);
+
+	REGISTER_GLOBAL(
+	    "CellPhoneLock", "dialogue_handler",
+	    "72 54 ? ? 5D ? ? ? 71 2C ? ? ? 2B 72 54 ? ? 77 54 ? ? 52 ? ? 25 ? 2C ? ? ? 53 ? ? 06 56 ? ? 52 ? ? 76", 2,
+	    GlobalPatternIdiom::GLOBAL_U16);
```

Comment:
Since we are calling those on-demand now it might be better design-wise to instead return a handle object which can then be passed to GlobalExists and GetGlobalAddr. These handles can then be cached in the effect itself so we don't have to worry about implementing some caching mechanism in the component itself.